FIPS/fip-draft_foundational_fvm.md

```diff
@@ -0,0 +1,1010 @@
+    fn get_chain_randomness_looking_forward(
+        &self,
+        pers: DomainSeparationTag,
+        round: ChainEpoch,
+        entropy: &[u8],
+    ) -> Result<[u8; 32]>;
+
+    /// Gets 32 bytes of randomness for ChainRand paramaterized by the DomainSeparationTag,
+    /// ChainEpoch, Entropy from the latest beacon entry.
+    fn get_beacon_randomness(
+        &self,
+        pers: DomainSeparationTag,
+        round: ChainEpoch,
+        entropy: &[u8],
+    ) -> Result<[u8; 32]>;
+
+    fn get_beacon_randomness_looking_forward(
+        &self,
+        pers: DomainSeparationTag,
+        round: ChainEpoch,
+        entropy: &[u8],
+    ) -> Result<[u8; 32]>;
+}
+
+```
+
+### IPLD memory model
+
+IPLD is the data model of Filecoin. Actors can be construed as logic that
+receives an input IPLD graph, performs computation, and (optionally) mutates the
+IPLD graph by saving an alternative root in the actor's state.
+
+All state will be stored and retrieved as IPLD. The state store lives on the
+node side, and it is exposed to the FVM via an Extern. This is important as the
+VM itself needs to understand the IPLD so we can do garbage collection, etc.
+
+IPLD codecs will be defined as WASM libraries, imported through the WASM import
+system. Ideally, ADLs and selectors would be implemented in the same way.
+
+The FVM must guarantee that actors can only access data that is in their state
+tree. This is done through the maintenance of an "accessible set" inside the
+kernel.
+
+Because Filecoin IPLD state objects are highly atomized and linked, accessing
+and mutating entries in objects like HAMTs and AMT results in multiple
+sequential state IO operations, each of which traverses the Extern boundary in a
+non-parallelizable way. If the Extern is traversed through FFI, the cost of
+operating on ADLs may be non-negligible.
+
+Alternative strategies like moving ownership of the state store to the FVM-side,
+or leveraging selectors for queries, may be explored in the future.
+
+### Error conditions
+
+There are two constructs in the FVM to signal results:
+
+- **Exit codes:** signal the outcome of the execution of an actor. They can be
+  emitted by the actor or the runtime, and they're recorded on-chain inside the
+  receipt.
+- **Syscall error numbers:** signal the outcome of a single syscall. They are
+  ephemeral, and have no meaning beyond the FVM syscall boundary.
+
+In both cases, a `0` value implies success. For more information, refer to the
+[FVM Error Conditions] specification (commit `724650` at the time of writing
+this FIP).
+
+### Memory policy
+
+The Invocation Container makes available a maximum of 1024 pages to the actor.
+Given that WASM memory pages are a fixed 64KiB size, this results in every actor
+call being limited to 64MiB of memory.
+
+Depending on the WASM engine, it may be possible to size memory dynamically. If
+that's the case, a policy that grows memory dynamically from a low value (even
+zero) should be preferred.
```

Comment:
I'd remove the details here as they're far from settled (and there are a lot of nuances involving wasm stack frames/depth, globals, locals, etc.).

The memory limit will also have to apply to the entire message execution. Otherwise the maximum would be max_depth * limit which is 64*4096 (I think?).

Reply:
I'll remove. We need to keep a holding pen document with unspecified details to figure out, though.